### PR TITLE
Unit Tests for Gaussian

### PR DESCRIPTION
**Pull request by jan-janssen**

### CODE REVIEW

File: tests/gaussian/test_gaussian.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                "n_alpha_electrons",
+                "n_basis_functions",
+                "n_beta_electrons",
+                "n_electrons",
+                "scf_density",
+                "spin_scf_density",
+            ]
+            with self.job_complete.project_hdf5.open("output/structure/dft") as h_gen:
+                hdf_nodes = h_gen.list_nodes()
+                self.assertTrue(all([node in hdf_nodes for node in nodes]))
```
Comment (jan-janssen) [inline tests/gaussian/test_gaussian.py]: I have the feeling the tests are not executed by our automated testing system, can you add: 

```
if __name__ == "__main__":
    unittest.main()
```

At the very end of the test file? Currently I expect the tests to fail, as the requirements are not installed in the test environment.

Comment (jan-janssen) [reply tests/gaussian/test_gaussian.py]: I already collected the requirements in https://github.com/pyiron/pyiron/pull/548

**Comment by jan-janssen** [comment]
It seems like the main method did not help. Can you try to add an empty `__init__.py` file to the test directory?

**Comment by jan-janssen** [comment]
Now the tests are executed and fail with: 

```
======================================================================
2258ERROR: test_run_bsse_complete (gaussian.test_gaussian.TestGaussian)
2259----------------------------------------------------------------------
2260Traceback (most recent call last):
2261  File "/home/travis/build/pyiron/pyiron/tests/gaussian/test_gaussian.py", line 150, in test_run_bsse_complete
2262    posixpath.join(self.execution_path, "../static/gaussian_test_files/bsse/input.log")
2263  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 479, in read
2264    parallel=parallel, **kwargs))
2265  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 537, in _iread
2266    for dct in io.read(fd, *args, **kwargs):
2267  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 340, in wrap_read_function
2268    yield read(filename, **kwargs)
2269  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/gaussian.py", line 112, in read_gaussian_out
2270    atoms.set_calculator(calc)
2271  File "/home/travis/build/pyiron/pyiron/pyiron/atomistics/structure/atoms.py", line 2769, in __getattr__
2272    return object.__getattribute__(self, item)
2273AttributeError: 'Atoms' object has no attribute 'set_calculator'
```

**Comment by SanderBorgmans** [comment]
> Now the tests are executed and fail with:
> 
> ```
> ======================================================================
> 2258ERROR: test_run_bsse_complete (gaussian.test_gaussian.TestGaussian)
> 2259----------------------------------------------------------------------
> 2260Traceback (most recent call last):
> 2261  File "/home/travis/build/pyiron/pyiron/tests/gaussian/test_gaussian.py", line 150, in test_run_bsse_complete
> 2262    posixpath.join(self.execution_path, "../static/gaussian_test_files/bsse/input.log")
> 2263  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 479, in read
> 2264    parallel=parallel, **kwargs))
> 2265  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 537, in _iread
> 2266    for dct in io.read(fd, *args, **kwargs):
> 2267  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/formats.py", line 340, in wrap_read_function
> 2268    yield read(filename, **kwargs)
> 2269  File "/home/travis/miniconda/lib/python3.7/site-packages/ase/io/gaussian.py", line 112, in read_gaussian_out
> 2270    atoms.set_calculator(calc)
> 2271  File "/home/travis/build/pyiron/pyiron/pyiron/atomistics/structure/atoms.py", line 2769, in __getattr__
> 2272    return object.__getattribute__(self, item)
> 2273AttributeError: 'Atoms' object has no attribute 'set_calculator'
> ```

This was the error I encountered too, which I fixed by changing the order of imports. But I guess that does not work here.

**Comment by jan-janssen** [comment]
Maybe it is easier to use the read input function from ASE directly: 
https://gitlab.com/ase/ase/-/blob/master/ase/io/gaussian.py#L230 ?

**Comment by SanderBorgmans** [comment]
> Maybe it is easier to use the read input function from ASE directly:
> https://gitlab.com/ase/ase/-/blob/master/ase/io/gaussian.py#L230 ?

I think this will throw the same error, but we can try. Moreover, this loads the input structure instead of the last frame of the log file, that will be important if tests are included that test the geometry.

**Comment by jan-janssen** [comment]
There is also a function to parse the output: 
https://gitlab.com/ase/ase/-/blob/master/ase/io/gaussian.py#L300 

The new error message: 

```
======================================================================
1338ERROR: test_run_sp_complete (gaussian.test_gaussian.TestGaussian)
1339----------------------------------------------------------------------
1340Traceback (most recent call last):
1341  File "C:\projects\pyiron\tests\gaussian\test_gaussian.py", line 100, in test_run_sp_complete
1342    posixpath.join(self.execution_path, "../static/gaussian_test_files/sp/input.com")
1343  File "C:\Miniconda37-x64\lib\site-packages\ase\io\gaussian.py", line 146, in read_gaussian
1344    atoms += Atom(symbol, position=position)
1345  File "C:\projects\pyiron\pyiron\atomistics\structure\atoms.py", line 2675, in __add__
1346    return self + other
1347  File "C:\projects\pyiron\pyiron\atomistics\structure\atoms.py", line 2642, in __add__
1348    sum_atoms._tag_list = sum_atoms._tag_list + other._tag_list
1349  File "C:\projects\pyiron\pyiron\atomistics\structure\sparse_list.py", line 509, in __add__
1350    new_length = len(self) + len(other)
1351TypeError: 'NoneType' object cannot be interpreted as an integer
1352
1353----------------------------------------------------------------------
1354Ran 430 tests in 260.123s
1355
1356
```

**Comment by SanderBorgmans** [comment]
I think the errors still have to do with the Atoms class being overloaded. Maybe we could try clearing all imports in the test_gaussian.py module?

**Comment by jan-janssen** [comment]
Can you try merging with https://github.com/pyiron/pyiron/pull/556 ?

**Comment by jan-janssen** [comment]
Ok, now we come to the point where it is missing your packages: 

```
======================================================================
2595ERROR: test_run_bsse_complete (gaussian.test_gaussian.TestGaussian)
2596----------------------------------------------------------------------
2597Traceback (most recent call last):
2598  File "/home/travis/build/pyiron/pyiron/tests/gaussian/test_gaussian.py", line 169, in test_run_bsse_complete
2599    self.job_complete.run()
2600  File "/home/travis/build/pyiron/pyiron/pyiron/base/job/generic.py", line 679, in run
2601    self._run_if_collect()
2602  File "/home/travis/build/pyiron/pyiron/pyiron/base/job/generic.py", line 1396, in _run_if_collect
2603    self.collect_output()
2604  File "/home/travis/build/pyiron/pyiron/pyiron/gaussian/gaussian.py", line 60, in collect_output
2605    output_dict = collect_output(output_file=os.path.join(self.working_directory, 'input.fchk'))
2606  File "/home/travis/build/pyiron/pyiron/pyiron/gaussian/gaussian.py", line 597, in collect_output
2607    fchk = FCHKFile(output_file)
2608NameError: name 'FCHKFile' is not defined
```

https://github.com/pyiron/pyiron/pull/554#discussion_r393750432

**Comment by SanderBorgmans** [comment]
> Ok, now we come to the point where it is missing your packages:
> 
> ```
> ======================================================================
> 2595ERROR: test_run_bsse_complete (gaussian.test_gaussian.TestGaussian)
> 2596----------------------------------------------------------------------
> 2597Traceback (most recent call last):
> 2598  File "/home/travis/build/pyiron/pyiron/tests/gaussian/test_gaussian.py", line 169, in test_run_bsse_complete
> 2599    self.job_complete.run()
> 2600  File "/home/travis/build/pyiron/pyiron/pyiron/base/job/generic.py", line 679, in run
> 2601    self._run_if_collect()
> 2602  File "/home/travis/build/pyiron/pyiron/pyiron/base/job/generic.py", line 1396, in _run_if_collect
> 2603    self.collect_output()
> 2604  File "/home/travis/build/pyiron/pyiron/pyiron/gaussian/gaussian.py", line 60, in collect_output
> 2605    output_dict = collect_output(output_file=os.path.join(self.working_directory, 'input.fchk'))
> 2606  File "/home/travis/build/pyiron/pyiron/pyiron/gaussian/gaussian.py", line 597, in collect_output
> 2607    fchk = FCHKFile(output_file)
> 2608NameError: name 'FCHKFile' is not defined
> ```
> 
> [#554 (comment)](https://github.com/pyiron/pyiron/pull/554#discussion_r393750432)

I thought the molmod and tamkin packages were already included?

**Comment by jan-janssen** [comment]
Can you merge with https://github.com/pyiron/pyiron/pull/548/files ?

**Comment by jan-janssen** [comment]
The issue is that molmod is not yet Python 3.8 ready https://github.com/molmod/molmod/pull/33

**Comment by jan-janssen** [comment]
Perfect - Here you can see which lines are covered by the test https://coveralls.io/jobs/60266832/source_files/2016157064 - the coveralls bot should update his post accordingly soon.

**Comment by jan-janssen** [comment]
The Azure tests fail because of the Python 3.8 issue in molmod, apart from this the tests are ready to be merged - https://github.com/molmod/molmod/pull/33

**Comment by jan-janssen** [comment]
@SanderBorgmans You have to be very careful with removing commits https://github.com/pyiron/pyiron/pull/554/commits/73dd71b89ad0da31a92b71c54eea3bbb426cffdc - the issue is if we merge both pull requests the NMA module is going to be removed, because these are the latest changes.

**Comment by SanderBorgmans** [comment]
I see what you mean. What is the best way to fix this, keeping the modules separate?

**Comment by jan-janssen** [comment]
I would suggest that we just start merging the changes and in the worst case you simply open another pull request for the NMA module. I just wanted to mention the issue here before it results in confusion later on. Btw. do you have a minute for a quick Skype call?

**Comment by SanderBorgmans** [comment]
@jan-janssen I understand. I have some time now, if you are free for a Skype call.

**Comment by jan-janssen** [comment]
I just was off for dinner - I am available now. If you can not reach me, just email me.

**Pull request merged**